File: android/app/src/main/java/com/rafaeltmbr/stopwatch/platform/presentation/view_models/LapsViewModel.kt
package com.rafaeltmbr.stopwatch.platform.presentation.view_models

import com.rafaeltmbr.stopwatch.core.entities.Status
import com.rafaeltmbr.stopwatch.platform.presentation.entities.ViewLap
import kotlinx.coroutines.flow.StateFlow


interface LapsViewModel {
    data class State(
        val status: Status = Status.INITIAL,
        val time: String = "00:00.00",
        val lapsCount: Int = 0
    )

    sealed class Action {
        data object Resume : Action()
        data object Lap : Action()
        data object NavigateBack : Action()
    }

    val state: StateFlow<State>
    fun handleAction(action: Action)
    fun getViewLapByReversedArrayIndex(index: Int): ViewLap
}

File: ios/Stopwatch/platform/presentation/view_models/LapsViewModel.swift
import Foundation

struct LapsState {
    let status: Status
    let milliseconds: String
    let lapsCount: Int
    
    init(status: Status = .initial, milliseconds: String = "00:00.00", lapsCount: Int = 0) {
        self.status = status
        self.milliseconds = milliseconds
        self.lapsCount = lapsCount
    }
}

enum LapsAction {
    case resume, lap
}

protocol LapsViewModel: ObservableObject {
    var state: LapsState { get }
    
    func handleAction(_ action: LapsAction)
    func getViewLapByReversedArrayIndex(_ index: Int) -> ViewLap
}
